src/wasm/wasm-type.cpp

```diff
@@ -55,46 +97,64 @@ namespace {
 
 std::mutex mutex;
 
-std::array<std::vector<Type>, Type::_last_value_type + 1> basicTypes = {
-  {{},
-   {Type::unreachable},
-   {Type::i32},
-   {Type::i64},
-   {Type::f32},
-   {Type::f64},
-   {Type::v128},
-   {Type::funcref},
-   {Type::externref},
-   {Type::nullref},
-   {Type::exnref}}};
+// Maps basic types to tuples of the single basic type.
+std::array<TypeList, Type::_last_basic_id + 1> basicTuples = {{
+  {},
+  {Type::unreachable},
+  {Type::i32},
+  {Type::i64},
+  {Type::f32},
+  {Type::f64},
+  {Type::v128},
+  {Type::funcref},
+  {Type::externref},
+  {Type::nullref},
+  {Type::exnref},
+}};
 
 // Track unique_ptrs for constructed types to avoid leaks
-std::vector<std::unique_ptr<std::vector<Type>>> constructedTypes;
-
-// Maps from type vectors to the canonical Type ID
-std::unordered_map<std::vector<Type>, uintptr_t> indices = {
-  {{}, Type::none},
-  {{Type::unreachable}, Type::unreachable},
-  {{Type::i32}, Type::i32},
-  {{Type::i64}, Type::i64},
-  {{Type::f32}, Type::f32},
-  {{Type::f64}, Type::f64},
-  {{Type::v128}, Type::v128},
-  {{Type::funcref}, Type::funcref},
-  {{Type::externref}, Type::externref},
-  {{Type::nullref}, Type::nullref},
-  {{Type::exnref}, Type::exnref},
+std::vector<std::unique_ptr<TypeDef>> constructedTypes;
+
+// Maps from constructed types to the canonical Type ID.
+std::unordered_map<TypeDef, uintptr_t> indices = {
+  {TypeDef(Tuple()), Type::none},
+  {TypeDef({Type::unreachable}), Type::unreachable},
+  {TypeDef({Type::i32}), Type::i32},
+  {TypeDef({Type::i64}), Type::i64},
+  {TypeDef({Type::f32}), Type::f32},
+  {TypeDef({Type::f64}), Type::f64},
+  {TypeDef({Type::v128}), Type::v128},
+  {TypeDef({Type::funcref}), Type::funcref},
+  {TypeDef({Type::externref}), Type::externref},
+  {TypeDef({Type::nullref}), Type::nullref},
+  {TypeDef({Type::exnref}), Type::exnref},
```

Comment:
I had the impression that these are there to avoid getting actual constructed Tuple types of a single type. Given a tuple of a single type, one will get the canonical `BasicID` of the single type, with the `TypeDef`s defined here being just placeholders (wrapping the respective vectors) to enable the lookup (but not used otherwise)?

Reply:
Oh hmm, good point. I suppose we have a lot of places in the code where we construct singleton (or empty) vectors and pass them to the `Type` constructor and expect to get a canonical type back, and changing these keys would make that no longer work. It would be good to add a comment to the code about why we are using tuples here.